File: tutorials/004_model_options_chapter/model_options_chapter.cpp
#include "neat.h"

#include <opencv2/core.hpp>

#include <array>
#include <filesystem>
#include <iostream>
#include <string>
#include <cstdlib>
#include <exception>
#include <initializer_list>
#include <stdexcept>
#include <utility>
#include <vector>

namespace {

bool has_flag(int argc, char** argv, const std::string& key) {
  for (int i = 1; i < argc; ++i) {
    if (key == argv[i]) return true;
  }
  return false;
}

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

bool wants_help(int argc, char** argv) {
  return has_flag(argc, argv, "--help") || has_flag(argc, argv, "-h");
}

bool wants_print_gst(int argc, char** argv) {
  return has_flag(argc, argv, "--print-gst");
}

void print_common_flags(std::ostream& os) {
  os << "  --help               Show this help message\n";
  os << "  --print-gst          Print the gst-launch string and exit\n";
}

int skip(const std::string& reason) {
  std::cout << "SKIP: " << reason << "\n";
  return 0;
}

bool strict_mode() {
  return std::getenv("SIMA_RUN_TUTORIALS_FULL") != nullptr;
}

std::string yes_no(bool v) { return v ? "yes" : "no"; }

void step(const std::string& name, const std::string& detail = {}) {
  if (detail.empty()) {
    std::cout << "STEP " << name << "\n";
  } else {
    std::cout << "STEP " << name << ": " << detail << "\n";
  }
}

void check(const std::string& name, bool cond, const std::string& detail = {}) {
  std::cout << "CHECK " << name << ": " << (cond ? "PASS" : "FAIL");
  if (!detail.empty()) std::cout << " (" << detail << ")";
  std::cout << "\n";
  if (!cond) throw std::runtime_error("check failed: " + name);
}

void why(const std::string& detail) {
  std::cout << "WHY " << detail << "\n";
}

void tradeoff(const std::string& detail) {
  std::cout << "TRADEOFF " << detail << "\n";
}

void failure_mode(const std::string& detail) {
  std::cout << "FAILURE_MODE " << detail << "\n";
}

void interpret_output(const std::string& detail) {
  std::cout << "INTERPRET " << detail << "\n";
}

void runtime_fallback(const std::exception& e) {
  const char* what = e.what();
  std::cout << "runtime_fallback: " << (what ? what : "unknown") << "\n";
}

void print_signature(std::initializer_list<std::pair<std::string, std::string>> values) {
  static constexpr std::array<const char*, 7> kRequired = {
      "tutorial", "lang", "flow", "run_mode", "output_kind", "tensor_rank", "field_count",
  };
  for (const char* key : kRequired) {
    bool found = false;
    for (const auto& kv : values) {
      if (kv.first == key) { found = true; break; }
    }
    if (!found) throw std::invalid_argument(std::string("missing signature key: ") + key);
  }
  std::cout << "SIGNATURE {";
  bool first = true;
  for (const auto& kv : values) {
    if (!first) std::cout << ",";
    std::cout << kv.first << "=" << kv.second;
    first = false;
  }
  std::cout << "}\n";
}

std::filesystem::path find_repo_root() {
  namespace fs = std::filesystem;
  fs::path cur = fs::current_path();
  for (int i = 0; i < 6; ++i) {
    if (fs::exists(cur / "CMakeLists.txt") && fs::exists(cur / "include") &&
        fs::exists(cur / "tests")) {
      return cur;
    }
    if (!cur.has_parent_path()) break;
    cur = cur.parent_path();
  }
  return fs::current_path();
}

std::filesystem::path first_existing(std::initializer_list<std::filesystem::path> candidates) {
  for (const auto& c : candidates) {
    if (std::filesystem::exists(c)) return c;
  }
  return {};
}

std::filesystem::path default_yolo_mpk() {
  namespace fs = std::filesystem;
  const fs::path root = find_repo_root();
  return first_existing({
      root / "tmp" / "yolo_v8s_mpk.tar.gz",
      root / "tmp" / "yolov8s_mpk.tar.gz",
      root / "tmp" / "yolo_mpk.tar.gz",
  });
}

std::filesystem::path default_resnet_mpk() {
  namespace fs = std::filesystem;
  const fs::path root = find_repo_root();
  return first_existing({
      root / "tmp" / "resnet_50_mpk.tar.gz",
      root / "tmp" / "resnet50_mpk.tar.gz",
  });
}

} // namespace

namespace fs = std::filesystem;

namespace {

void print_help(const char* argv0) {
  std::cout << "Usage: " << argv0 << " [--mpk <path>]\n";
  print_common_flags(std::cout);
  std::cout << "  --mpk <path>         Path to model MPK\n";
}

void print_spec(const char* label, const simaai::neat::TensorConstraint& spec) {
  std::cout << label << ": rank=" << spec.rank << " dtypes=" << spec.dtypes.size()
            << " shape_dims=" << spec.shape.size() << "\n";
}

} // namespace

int main(int argc, char** argv) {
  try {
    if (wants_help(argc, argv)) {
      print_help(argv[0]);
      return 0;
    }

    // Why: explicit runtime markers keep the tutorial explainable from terminal output alone.
    // Why: parity and score tooling consume these checkpoints as a stable contract.
    step("input_contract", "parse flags and establish deterministic defaults");
    step("run_mode_choice", "exercise the chapter's primary runtime path");
    why("understand the contract first: inputs, run mode, and outputs");
    tradeoff(
        "prefer deterministic samples and stable contracts over production realism");
    failure_mode(
        "runtime/plugin issues should degrade to runtime_fallback without losing observability");
    interpret_output(
        "use CHECK markers plus SIGNATURE fields to validate behavior and parity");
    step("output_contract", "emit checks and machine-parseable signature");
    check("strict_flag_available",
                       yes_no(strict_mode()) == "yes" ||
                           yes_no(strict_mode()) == "no",
                       "strict-mode guard is observable");

    const fs::path root = find_repo_root();

    std::string mpk_arg;
    fs::path mpk_path = get_arg(argc, argv, "--mpk", mpk_arg)
                            ? fs::path(mpk_arg)
                            : first_existing({
                                  default_yolo_mpk(),
                                  default_resnet_mpk(),
                              });
    if (mpk_path.empty() || !fs::exists(mpk_path)) {
      return skip("missing MPK (pass --mpk)");
    }

    // CORE LOGIC
    simaai::neat::Model::Options opt;
    opt.media_type = "video/x-raw";
    opt.format = "BGR";
    opt.input_max_width = 640;
    opt.input_max_height = 640;
    opt.input_max_depth = 3;
    opt.preproc.normalize = true;
    opt.preproc.channel_mean = std::array<float, 3>{0.485f, 0.456f, 0.406f};
    opt.preproc.channel_stddev = std::array<float, 3>{0.229f, 0.224f, 0.225f};
    opt.decode_type = "yolov8";
    opt.score_threshold = 0.35f;
    opt.nms_iou_threshold = 0.45f;
    opt.top_k = 100;
    opt.original_width = 640;
    opt.original_height = 640;
    opt.name_suffix = "_chapter";

    simaai::neat::Model model(mpk_path.string(), opt);
    // END CORE LOGIC

    print_spec("input_spec", model.input_spec());
    print_spec("output_spec", model.output_spec());
    std::cout << "metadata keys: " << model.metadata().size() << "\n";

    if (wants_print_gst(argc, argv)) {
      simaai::neat::Session s;
      s.add(model.session());
      std::cout << s.describe_backend() << "\n";
      return 0;
    }

    cv::Mat bgr(224, 224, CV_8UC3, cv::Scalar(10, 20, 30));
    if (!bgr.isContinuous()) {
      bgr = bgr.clone();
    }

    try {
      auto out = model.run(bgr, 2000);
      std::cout << "run() output kind: " << static_cast<int>(out.kind) << "\n";
    } catch (const std::exception& e) {
      // Deterministic fallback keeps strict runs pedagogically useful when device plugins
      // misconfigure.
      runtime_fallback(e);
    }

    check("tutorial_completed", true, "main path reached end without exception");
    print_signature({
        {"tutorial", "004"},
        {"lang", "cpp"},
        {"flow", "chapter_path"},
        {"run_mode", "sync_or_async"},
        {"output_kind", "sample_or_tensor"},
        {"tensor_rank", "-1"},
        {"field_count", "-1"},
    });

    std::cout << "[OK] 004_model_options_chapter\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/004_model_options_chapter/model_options_chapter.py
#!/usr/bin/env python3
from __future__ import annotations

import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit("pyneat is not importable. Either NEAT is not installed, or the venv is not activated.\nRun: source ~/pyneat/bin/activate\nIf the venv does not exist yet, follow the installation guide.")

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "common"))
import python_utils as tu


def main(argv: list[str]) -> int:
  import numpy as np

  if tu.has_flag(argv, "--help"):
    print(f"Usage: {argv[0]} [--mpk <path>] [--print-gst]")
    return 0

  # Why: runtime markers make intent explicit without requiring external docs.
  # Why: parity/scorecard tooling relies on stable, machine-parseable checkpoints.
  tu.step("input_contract", "parse flags and establish deterministic defaults")
  tu.step("run_mode_choice", "exercise the chapter's primary runtime path")
  tu.why("understand the contract first: inputs, run mode, and outputs")
  tu.tradeoff("prefer deterministic samples and stable contracts over production realism")
  tu.failure_mode("runtime/plugin issues should degrade to runtime_fallback without losing observability")
  tu.interpret_output("use CHECK markers plus SIGNATURE fields to validate behavior and parity")
  tu.step("output_contract", "emit checks and machine-parseable signature")
  tu.check("strict_flag_available", isinstance(tu.strict_mode(), bool),
           "strict-mode guard is observable")

  root = tu.repo_root()
  mpk_arg = tu.get_arg(argv, "--mpk")
  mpk = Path(mpk_arg) if mpk_arg else tu.first_existing([tu.default_yolo_mpk(root), tu.default_resnet_mpk(root)])
  if not mpk or not mpk.exists():
    return tu.skip("missing MPK (pass --mpk)")

  # CORE LOGIC
  opt = pyneat.ModelOptions()
  opt.format = "BGR"
  opt.input_max_width = 640
  opt.input_max_height = 640
  opt.input_max_depth = 3
  opt.decode_type = "yolov8"
  opt.score_threshold = 0.35
  opt.nms_iou_threshold = 0.45
  opt.top_k = 100
  opt.original_width = 640
  opt.original_height = 640
  opt.name_suffix = "_chapter"

  opt.preproc.normalize = True
  opt.preproc.channel_mean = [0.485, 0.456, 0.406]
  opt.preproc.channel_stddev = [0.229, 0.224, 0.225]

  model = pyneat.Model(str(mpk), opt)
  # END CORE LOGIC
  print(f"input_spec.rank:  {model.input_spec().rank}")
  print(f"output_spec.rank: {model.output_spec().rank}")
  print(f"metadata keys:    {len(model.metadata())}")

  if tu.has_flag(argv, "--print-gst"):
    s = pyneat.Session()
    s.add(model.session())
    print(s.describe_backend())
    return 0

  rgb = np.full((224, 224, 3), 44, dtype=np.uint8)
  t = pyneat.Tensor.from_numpy(rgb, copy=True, image_format=pyneat.PixelFormat.BGR)
  try:
    out = model.run(t, timeout_ms=2000)
    print(f"run() output kind: {out.kind}")
  except Exception as exc:
    # Deterministic fallback keeps strict runs pedagogically useful when device plugins misconfigure.
    tu.runtime_fallback(exc)

  tu.check("tutorial_completed", True, "main path reached end without exception")
  tu.signature({
      "tutorial": "004",
      "lang": "py",
      "flow": "chapter_path",
      "run_mode": "sync_or_async",
      "output_kind": "sample_or_tensor",
      "tensor_rank": -1,
      "field_count": -1,
  })

  print("[OK] 004_model_options_chapter")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
